### PR TITLE
[processor/k8sattributes] Add first version of k8s e2e tests

### LINKED ISSUE #15651: Introduce e2e tests for kubernetes story
## Problem statement

Kubernetes related components, especially [k8s attributes processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/k8sattributesprocessor) are used pretty extensively in the field. At the same time they don't have good e2e test coverage. Recently we had a significant change that introduced a few regressions. We want to make sure that such regressions are caught by CICD going forward.

## Proposed solution

We need to introduce a GitHub actions job that would spin up a k8s cluster with a collector built from the current branch. 

Proposed steps:

1. Build a collector docker image with required k8s related components.
2. Spin up a [kind k8s cluster](https://kind.sigs.k8s.io/docs/user/quick-start/).
3. Spin up a collector with applications generating a predictable load: metrics, logs, and traces. We can use https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-demo. The collector should be configured with [file exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/fileexporter)
4. The OTLP JSON files written by the collector should be parsed and compared against expected values, e.g. k8s pod/deployment/namespace UIDs/names/labels/annotations etc.

## Action items

- [x] Spike to verify (or change) the proposed solution
- [x] e2e test for k8s attributes processor: traces
- [x] #18514
- [x] #18391
- [x] #18392
- [x] #18393
- [x] #18394
- [x] #18395

Once it's introduced in the contrib repo, we can potentially reuse the same tests in the helm repo.

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

**Description:** 
Updates #15651

**Link to tracking Issue:** #15651

### CODE REVIEW

**Comment by dmitryax** [comment]
@fatsheep9146 please run `make crosslink` and `make gotidy` to make the checks job pass and fix the linter issues

**Comment by dmitryax** [comment]
This is a great start! Let's keep https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/15651 open until we have coverage of all k8s components

File: processor/k8sattributesprocessor/e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		len := traces.ResourceSpans().Len()
+		for i := 0; i < len; i++ {
+			resourceSpans := traces.ResourceSpans().At(i)
+
+			t.Logf("resource attributes: %s", resourceSpans.Resource().Attributes().AsRaw())
+
+			resourceSpans.Resource().Attributes().Range(
+				func(k string, v pcommon.Value) bool {
+					if k == "k8s.job.name" && v.Str() == "test-telemetrygen" {
+						jobKeyFound = true
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:68]: Should we count here to make sure that every span has the keys? `jobAttrCount++`

File: processor/k8sattributesprocessor/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	nsKeyFound := false
+	for scanner.Scan() {
+		traces, err := unmarshaler.UnmarshalTraces(scanner.Bytes())
+		assert.NoError(t, err)
+
+		len := traces.ResourceSpans().Len()
+		for i := 0; i < len; i++ {
+			resourceSpans := traces.ResourceSpans().At(i)
+
+			t.Logf("resource attributes: %s", resourceSpans.Resource().Attributes().AsRaw())
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:63]: Please remove the debug line

File: processor/k8sattributesprocessor/e2e_test.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+					if k == "k8s.namespace.name" && v.Str() == "default" {
+						nsKeyFound = true
+					}
+					return true
+				},
+			)
+		}
+	}
+
+	assert.Equal(t, true, jobKeyFound && nsKeyFound, "job and namespace key are not found")
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:79]: Then we would be

```suggestion
	require.Less(t, 0, totalResourceSpans)
	assert.Equal(t, totalResourceSpans, jobAttrCount)
	assert.Equal(t, totalResourceSpans, nsAttrCount)
```

File: processor/k8sattributesprocessor/e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	if c == io.EOF {
+		return nil, true, nil
+	}
+	return buf, false, nil
+}
+
+func TestJobE2E(t *testing.T) {
+	if _, err := os.Stat("./testdata/trace.json"); errors.Is(err, os.ErrNotExist) {
+		return
+	}
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:40]: Not needed

File: processor/k8sattributesprocessor/e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+func readJSONMessage(br *bufio.Reader) ([]byte, bool, error) {
+	buf, _, c := br.ReadLine()
+	if c == io.EOF {
+		return nil, true, nil
+	}
+	return buf, false, nil
+}
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:35]: Unused

File: .github/workflows/k8s-related-components-e2e-collector-values.yml [hunk 6]
```diff
@@ -0,0 +1,416 @@
+# Default values for opentelemetry-collector.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
```
Comment (dmitryax) [inline .github/workflows/k8s-related-components-e2e-collector-values.yml:3]: Please remove everything from this file expect for the fields that are overridden

**Comment by dmitryax** [comment]
@fatsheep9146 please let me know if you can resolve the comment or you need help

**Comment by fatsheep9146** [comment]
> opentelemetry

I will start to fix your comments @dmitryax

File: .github/workflows/e2e/k8s/collector-helm-values.yml [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            static_configs:
+              - targets:
+                  - ${MY_POD_IP}:8888
+  service:
+    telemetry:
+      metrics:
+        address: ${MY_POD_IP}:8888
+    extensions:
+      - health_check
+      - memory_ballast
```
Comment (dmitryax) [inline .github/workflows/e2e/k8s/collector-helm-values.yml:39]: All these lines are defined in the default values.yaml. Let's not duplicate them here, but keep only overrides only

File: .github/workflows/e2e/k8s/collector-helm-values.yml [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+          - prometheus
+      traces:
+        exporters:
+          - logging
+          - file/trace
+        processors:
+          - memory_limiter
+          - batch
+        receivers:
+          - otlp
```
Comment (dmitryax) [inline .github/workflows/e2e/k8s/collector-helm-values.yml:67]: Let's define only tracing pipeline for now until we add tests for others
```suggestion
      traces:
        exporters:
          - file/trace
```

**Review by dmitryax — COMMENTED** [review]
LGTM. Let's clean up the values.yaml and merge this as the first version

**Comment by fatsheep9146** [comment]
Failed test seems to be flaky.

**Comment by dmitryax** [comment]
@fatsheep9146 can you add other data types to the test as a follow-up work?

**Review by dmitryax — APPROVED** [review]
Thank you, @fatsheep9146!

**Comment by fatsheep9146** [comment]
> @fatsheep9146 can you add other data types to the test as a follow-up work?

Absolutely! @dmitryax

**Pull request merged**